File: Data Structures & Algorithms/binary-tree-maximum-path-sum/submission-0.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */

class Solution {
    public int maxPathSum(TreeNode root) {
        int[] ans = new int[1];
         ans[0] = Integer.MIN_VALUE;
        dfs(root,ans);
        return ans[0];





       
        
    }

    public int dfs(TreeNode root, int[]res){
        if(root==null) return 0;

        int left = dfs(root.left,res);
        int right = dfs(root.right,res);
        left = Math.max(0,left);
        right = Math.max(0,right);
        res[0] = Math.max(res[0] , left+right+root.val);

        return root.val+Math.max(left,right);
    }
}

File: Data Structures & Algorithms/lowest-common-ancestor-in-binary-search-tree/submission-0.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right

class Solution:
    def lowestCommonAncestor(self, root: TreeNode, p: TreeNode, q: TreeNode) -> TreeNode:
        if root is None:
            return None  #empty tree
        #we found either p or q
        if root==p or root==q:
            return root

        #recursively call left subtree and right subtree
        leftresult=self.lowestCommonAncestor(root.left,p,q)
        rightresult=self.lowestCommonAncestor(root.right,p,q)

        #if both leftresult and rightresult found return root that particular one is te
        if leftresult is not None and rightresult is not None:
            return root


        #only if left reuslt found return 
        if leftresult is not None:
            return leftresult
        return rightresult
        